### ISSUE TITLE
Zlib compression needs to be sync

### ISSUE THREAD

**Issue by alexhultman** [websockets/ws#1369]
I've read and followed the zlib "memory fragmentation" issue down to where people blaming the Linux kernel and glibc and whatnot.

My two cents:

* There is nothing wrong with the Linux kernel
* There is nothing wrong with glibc
* There is nothing wrong with Node.js

The issue is with websockets/ws.

Doing CPU-bound tasks "async" is nothing more than a delusion - there is no such thing as an async CPU-bound task. If it takes 1 second of CPU-time, it takes 1 second of CPU time no matter when you perform it. Passing a CPU bound task to some thread pool doesn't improve anything in fact it only adds more overhead:

* Threading overhead
* You need to copy the data one extra time (major overhead!)
* You're basically just procrastinating what you need to do right now.

If you cannot keep up with what you buffer up, you're obviously going to blow the process up from too much memory usage since you produce more than you consume.

You have two solutions:

1. Replace all zlib calls with sync calls and all problems go away and you will lower latencies and actually do work you need to do and not just fool yourself you're doing things "in the background".

OR if you _really_ want to try and use multithreading here (which goes against how Node.js is built and deployed):

2. Only buffer up NUMBER_OF_LOGICAL_CPU_CORES with the async version and after this do sync calls until the buffer contains less than NUMBER_OF_LOGICAL_CPU_CORES tasks. Aka - you NEED to block sometime because CPU work cannot be "async" like networking. This solution is more complex and goes against what Node.js is but you get to choose.

**Comment by lpinca** [comment]
This issue https://github.com/nodejs/node/issues/8871#issuecomment-250915913 is the problem in ws and yes it will probably go away if sync calls are used but it's still valid as the async usage is what is actually recommended.

As written in the comment a different allocator can work around the issue and in fact, there is no leak on macOS.

I don't know if it's viable to use the Node.js zlib sync methods. It will certainly require extensive testing and big production environment to validate the results.

**Comment by alexhultman** [comment]
Not going to argue or repeat myself as I have no gain or need for doing so. I've already posted my hint and you may take it or reject it. Not my business to care for what you do.

All I know is the current websockets/ws crashes after three hours in production and cannot be used for this reason while uws does not, in the very same production environment and same OS. If you want to blame Linux that's your problem but like said uws has no problem working with zlib without constantly blowing up on Linux.

**Comment by alexhultman** [comment]
Honestly though, blaming glibc is hilariously ignorant in this case as it makes no sense from any standpoint to stubbornly claim "it is recommended" (by whom exactly!?) when it clearly is a one way ticket to horrible perf. and stability to just spam memory allocation after memory allocation just because you have this belief that "async" is actually a real thing here. Your'e basically just building an unbound mountain of unprocessed tasks without actually letting the CPU ever get to work and process the load. It's the most basic example of producer consumer except completely broken since you have no actual synchronization between the two parts (you just produce tasks without ever waiting for the consumer to finish).

glibc is used in every single program ever made for Linux for the last like 30 years but somehow it is "completely broken" because some script that spams 30000 copies of a task without waiting for them to finish is "recommended". The way Node.js exposes Zlib.deflate as this bullshit "async" wrapper of the well known zlib deflate is just mindblowingly stupid to begin with as it completely goes again every fiber of what Node.js is about.

But hey, what do I know?

**Comment by lpinca** [comment]
I'm not going to argue with you either. I've tried in the past, it's useless, there is no constructive discussion.
Thank you for the suggestion. If anyone want to experiment with sync zlib calls, cool. Please do that! Run benchmarks, use it in a heavy-loaded production environment, analyse the event loop lag, etc. Then eventually open a PR.

**Comment by alexhultman** [comment]
Okay, keep it broken and unstable then. Nobody can use permessage-deflate in production as is now since it constantly blows up. Using sync zlib would not "probably fix it", it would 100% guaranteed fix it. Not my problem.

**Comment by alexhultman** [comment]
We already did test it and uws does not blow up while websockets/ws does, every three hours. The difference is uws does it sync while websockets/ws just builds a mountain of unprocessed copies until it crashes.

**Comment by lpinca** [comment]
Where? BitMEX? In that case why isn't @STRML proposing a patch? They proposed multiple patches in the past to work around the issue. Or are they no longer interested now since they are using uws?

Open a PR, convince maintainers that it is the right thing to do (with numbers), PR merged, discussion ends, all users of ws benefit from this. How awesome is that.

**Comment by alexhultman** [comment]
Their patches never fixed the issue, it still blows up.

From a purely academic standpoint, what you are _trying_ to do by using "async" deflate is the well known "one producer many consumers" multithreading problem.

If you Google it up, you'll see that you _have_ to block the producer if the consumer queue is full. Problem is, JavaScript is not a multithreading language and does not expose threading fundamentals to solve it. That's why I'm telling you it "goes against everything Node.js is about" when trying to hack in proper threading by fooling users it is "async". You need to block, it is a very known academic hello world example taught in colleges.

**Comment by alexhultman** [comment]
By "full" I mean holding as many tasks as you have logical CPU cores. Which is like top 8 today.

**Comment by STRML** [comment]
Hey @lpinca - sorry for not updating you further. I'm happy to contribute a sync patch and put it through out benchmarks for completeness' sake. Even with concurrency=1 we still saw fragmentation-related blowups pretty much daily in production. I have not yet tested if sync will fix it but it rings true. Unfortunately this is a difficult bug to reproduce in test.

**Comment by alexhultman** [comment]
Blocking the event loop (with CPU work) is not the same as blocking it with sleep. Blocking with sleep aka "IO" is always bad but blocking with CPU work is always happening to some degree and inevitable.

**Comment by STRML** [comment]
Yeah it depends on how you've clustered out your app. If you're doing one app per hyperthread it's already pretty close to counterproductive to thread out zlib.

**Comment by alexhultman** [comment]
Yep any threading will add allocation, copy, synchronization and free overhead and since Node.js is all about being scaled one process per logical CPU core you only get added overhead which only turns the 1 second of required CPU time up, not down.

**Comment by lpinca** [comment]
> blocking with CPU work is always happening to some degree and inevitable.

Well this is why many computationally intensive tasks are delegated to the libuv thread pool no?

**Comment by STRML** [comment]
Well they can be. Zlib is a bit of an outlier because the standard task delegated to the libuv pool is IO, not CPU. If you assume that any process can actually spawn 1-5 CPU-intensive processes this completely breaks assumptions when clustering. Ideally you want to spawn one cluster worker per hyperthread otherwise.

**Comment by lpinca** [comment]
And here we are again with that arrogant attitude. My point is that it's not ws fault. Any Node.js app that uses the asynchronous zlib deflate method is potentially affected by that memory fragmentation issue.

Is it viable to use the sync methods without incurring in other issues? I don't know. Very happy if this is the case as the code will actually be also a lot easier to maintain.

**Comment by alexhultman** [comment]
It's not arrogant, it's just short and to the point. Remember: I'm here to _hint_ of an _improvement_ nothing else. It gets kind of tiresome just repeating myself over and over when I _know for a fact_ this to work.

> The point is that it's not ws fault.

Well okay then, ignore everything I've said in this thread and keep it broken.

**Comment by lpinca** [comment]
Issue should stay open, and I already said thank you for your suggestion, not sure what else should I do haha. Please explain me exactly what is ws doing wrong because I'm shortsighted. Is using a recommended and public API ws fault? Is it ws fault if that API leads to memory fragmentation?

**Comment by alexhultman** [comment]
Can't keep reiterating -> I'm off, good luck

**Comment by alexhultman** [comment]
Just a side note since I am banned in the nodejs repo:

I've spent like 6 hours of memory testing µWS in various permessage-deflate problems and really been hammering it in all kind of ways - I can definitely say that glibc works and is stable however it _does_ grow memory usage in a logarithmic way until it stagnates and goes flatline.

It is true that swapping to libtcmalloc as allocator makes µWS _complely_ flatlined in memory usage under immense permessage-deflate load, for hours, but the same can be said about glibc after a while when it has stagnated in memory usage.

It will definitely not continue to grow in a linear fashion like the async deflate example linked in this thread (have you tried running it over and over? It grows linearly!!). Even though glibc is far less likely to return memory to the system, it still manages to pick free blocks of memory from the already large heap (the bigger the heap, the lower the chance of not having a free block).

So it definitely _is_ possible to have a working and stable permessage-deflate implementation without any ever continuing memory issues, even with glibc.

**Comment by luislobo** [comment]
So, https://github.com/websockets/ws/pull/1204 does not fix the issue completely, right?

**Comment by lpinca** [comment]
@luislobo according to the above comment, yes that's right.

**Comment by rahbari** [comment]
After some testing I saw the async version of gzip takes up to 10 times more than the sync version according to the content size. 

**100KB:** sync version takes 5ms and async version 20ms
**2KB:** sync version takes 0.3ms and async version 3ms

As websocket usually is used for small chunk of data I think regardless of the memory leak it's better to use sync deflate considering under 1ms operations are neglectable.

I suggest adding another threshold option for sync compression.

**Comment by mk-pmb** [comment]
> usually is used for small chunk of json data

No need to speculate. We can set a limit for what's tiny enough to be compressed right away (i.e. sync).

**Comment by mk-pmb** [comment]
Sorry for not having read it properly a minute ago. :+1: for the threshold option.

**Comment by alexhultman** [comment]
> As websocket usually is used for small chunk of json data I think regardless of the memory leak it's better to use sync deflate for small data considering under 1ms operations are neglectable.

Bingo.

At these small messages (most users send less than 500 bytes a message) the threading, eventing, copying and synchronization overhead it means to "make it async" becomes a bigger cost than the actual compression itself. Making it "async" only fools oneself as it accomplishes nothing of value.

I would never put compression as "async" because it simply is logically impossible to do so, but I also don't care enough to argue anymore. Setting a threshold would solve it enough to be usable.

**Comment by piranna** [comment]
There should be two thresholds, one to enable compression and another to do
it async... Sometimes, gzip compressed data gets bigger than the original
one for small data chunks, and also there's the CPU usage for the
conpression.

**Comment by alexhultman** [comment]
I'm assuming you toss the compressed variant if it turns out bigger than the original.....

**Comment by mk-pmb** [comment]
I'll assume that for each length of input, gzip achives maximum efficiency if and only if all input bytes are the same. In this case, and using values from 0…127, the break-even is at 23 bytes.
Inputs shorter than 23 bytes will result in larger outputs.
Compression becomes efficient starting at 24 bytes of input.

**Comment by lpinca** [comment]
Node.js does not have an official API to use inflate/deflate streams synchronously.

We can't use https://nodejs.org/api/zlib.html#zlib_zlib_deflaterawsync_buffer_options as that does not work as intended for fragmented messages of if context takeover is enabled. We need the ability to reuse the same https://nodejs.org/api/zlib.html#zlib_class_zlib_deflateraw instance.

In order to use a deflate stream synchronously we need to copy this internal function https://github.com/nodejs/node/blob/5d210d4ac938a16d132a1433857927c5c58a8954/lib/zlib.js#L459-L527 but ofc that can change at any time.

As said in 
https://github.com/websockets/ws/issues/1369#issuecomment-383808666 experimentation (backed by numbers) is welcomed.

**Comment by alexhultman** [comment]
> As said in #1369 (comment) experimentation (backed by numbers) is welcomed.

See https://github.com/websockets/ws/issues/1369#issuecomment-414835835

**Comment by lpinca** [comment]
Yes, but where is the code that produced those results? How many deflate instances were used (we have one per socket)? How long is event loop blocked when doing that for say 50k clients? I would like to see this kind of stuff.

**Comment by alexhultman** [comment]
@lpinca See https://github.com/websockets/ws/issues/1369#issuecomment-383976309 ❤️

**Comment by lpinca** [comment]
> 2KB: sync version takes 0.3ms and async version 3ms

@rahbari can you please share your code?

I did a dummy and irrelevant quick test and this is what I get:

<details>
<summary>Code</summary>

```js
'use strict';

const crypto = require('crypto');
const zlib = require('zlib');

const buf = crypto.randomBytes(2048);

console.time('deflate sync');
for (let i = 0; i < 1000; i++) {
  zlib.deflateRawSync(buf);
}
console.timeEnd('deflate sync');

const deflate = zlib.createDeflateRaw();
let runs = 0;

function runAsync() {
  deflate.write(buf);
  deflate.resume();
  deflate.flush(() => {
    if (++runs === 1000) {
      console.timeEnd('deflate async');
      return;
    }

    runAsync();
  });
}

console.time('deflate async');
runAsync();
```
</details>

<details>
<summary>Results</summary>

```
$ for i in $(seq 1 10); do node index.js; echo ""; done
deflate sync: 148.891ms
deflate async: 169.522ms

deflate sync: 149.968ms
deflate async: 164.356ms

deflate sync: 150.329ms
deflate async: 164.349ms

deflate sync: 144.671ms
deflate async: 162.748ms

deflate sync: 163.011ms
deflate async: 164.543ms

deflate sync: 147.325ms
deflate async: 163.505ms

deflate sync: 172.986ms
deflate async: 169.949ms

deflate sync: 148.711ms
deflate async: 167.773ms

deflate sync: 146.679ms
deflate async: 168.316ms

deflate sync: 153.743ms
deflate async: 166.653ms

```
</details>

**Comment by piranna** [comment]
> I'll assume that for each length of input, gzip achives maximum efficiency if and only if all input bytes are the same. In this case, and using values from 0…127, the break-even is at 23 bytes.
> Inputs shorter than 23 bytes will result in larger outputs. For these, we shouldn't waste any CPU cycles on gzip and the length comparison.

Not sure about the 23 bytes number, but yes, that's exactly what I was talking about :-)

**Comment by piranna** [comment]
> Node.js does not have an official API to use inflate/deflate streams synchronously.

Streams not, but [there's for buffers](https://nodejs.org/api/zlib.html#zlib_convenience_methods).

**Comment by lpinca** [comment]
@piranna please read the full comment. That creates a new instance for each buffer, see https://github.com/nodejs/node/blob/5d210d4ac938a16d132a1433857927c5c58a8954/lib/zlib.js#L692-L706. We can't use that for fragmented messages or if context takeover is enabled.

**Comment by piranna** [comment]
But does compression being done at fragment level? Isn't it message level?

**Comment by lpinca** [comment]
It is done at the frame level yes.

**Comment by lpinca** [comment]
The option to enable compression based on the message byte size is available since version 2.0.0, see https://github.com/websockets/ws/blob/1e78aba9b9b29407517ec7ba1da97427c1678e5c/doc/ws.md#new-websocketserveroptions-callback

> `threshold` {Number} Payloads smaller than this will not be compressed. Defaults to 1024 bytes.

**Comment by alexhultman** [comment]
> Node.js does not have an official API to use inflate/deflate streams synchronously.

If that is the case then first step would be to add that feature in Node.js. Or use something like pako, I don't know.

**Comment by lpinca** [comment]
I noticed that there is [`Zlib.prototype._processChunk()`](https://github.com/nodejs/node/blob/e10290cc67edd5118d4d72efa5336dc2ba652d05/lib/zlib.js#L451-L457). The API is "private" and undocumented but it seems to work and it is used in user land so it will probably not go away anytime soon.

```js
const deflate = new zlib.DeflateRaw();
const data = deflate._processChunk(chunk, zlib.Z_SYNC_FLUSH);
```

**Comment by mk-pmb** [comment]
Would it help do deflate/inflate buffers instead of the stream? That way we could use zlib.deflateSync.

**Comment by fungiboletus** [comment]
The difficulty is to support "context takeover".

The RFC defines it like this :

> The term "LZ77 sliding window" used in this section means the
   buffer used by the DEFLATE algorithm to store recently processed
   input.  The DEFLATE compression algorithm searches the buffer for a
   match with the following input.

>   The term "use context takeover" used in this section means that the
   same LZ77 sliding window used by the endpoint to build frames of the
   previous sent message is reused to build frames of the next message
   to be sent.

In my opinion, context takeover is what makes websocket compression very cool. You can compress based on the previous messages, which is very convenient when you have many similar messages which I think is very common with websockets.

So, `ws` should keep the sliding window between messages and the easiest way to do it was to code something on top of zlib streams.

**Comment by lpinca** [comment]
@fungiboletus I can't reproduce on Ubuntu 18.04 using your example and Node.js 13.1.0:

<details>
<summary>server.js</summary>

```js
'use strict';

const WebSocket = require('ws');

const message = Buffer.alloc(256);

const wss = new WebSocket.Server({
  port: 8080,
  perMessageDeflate: {
    zlibDeflateOptions: {
      memLevel: 3
    },
    serverMaxWindowBits: 10,
    threshold: 16
  }
});

function onClose() {
  if (wss.clients.size === 0) {
    console.log('connections: 0');
  }
}

function broadcast() {
  console.log('Broadcasting message');

  for (const ws of wss.clients) {
    ws.send(message);
    ws.close();
  }
}

wss.on('connection', (ws) => {
  if (wss.clients.size === 1000) {
    console.log('connections: 1000');
    broadcast();
  }

  ws.on('close', onClose);
});

wss.on('listening', function() {
  console.log('Listening on *:8080');
});

setInterval(function() {
  console.log(process.memoryUsage());
}, 10000);
```
</details>

<details>
<summary>clients.js</summary>

```js
'use strict';

const WebSocket = require('ws');

const total = 1000;
let connected = 0;

function connect() {
  const ws = new WebSocket('ws://localhost:8080/');

  ws.on('open', function() {
    if (++connected < total) connect();

    ws.on('close', function() {
      if (--connected === 0) setTimeout(connect, 5000);
    });
  });
}

connect();
```
</details>

---

Memory usage stabilizes at ~100 MiB.

**Comment by LongTengDao** [comment]
Why not send and recieve compressed buffer in userland?

**Comment by chanced** [comment]
https://github.com/nodejs/node/pull/34048

**Comment by timotejroiko** [comment]
Greetings,

I am experimenting with synchronous zlib compression in `ws`.
This implementation is built on top of `Zlib.prototype._processChunk()` but because its also released as a standalone library, i added it to `ws` via a peer dependency to avoid tampering with the existing solution as much as possible. The code is still a bit rough and may be missing some error handling but here are some early test results made with a modified version of the `/bench/speed.js` test:

(windows 10 build 2004, node.js 14.8.0, intel i5 7300HQ 2.5ghz)

```bash
> node speedzlib.js
Generating 100 MiB of test data...
Testing ws on [::]:8181
10000 roundtrips of 64 B binary data:   0.9s    1.4 MiB/s
5000 roundtrips of 16 KiB binary data:  2.8s    55.45 MiB/s
1000 roundtrips of 128 KiB binary data: 13.5s   18.48 MiB/s
100 roundtrips of 1 MiB binary data:    9.6s    20.87 MiB/s
1 roundtrips of 100 MiB binary data:    9.4s    21.23 MiB/s
10000 roundtrips of 64 B text data:     0.9s    1.4 MiB/s
5000 roundtrips of 16 KiB text data:    2.6s    59.87 MiB/s
1000 roundtrips of 128 KiB text data:   13.2s   18.95 MiB/s
100 roundtrips of 1 MiB text data:      9.8s    20.34 MiB/s
1 roundtrips of 100 MiB text data:      9.5s    20.96 MiB/s

> npm install fast-zlib
> node speedzlib.js
Generating 100 MiB of test data...
Testing ws on [::]:8181
10000 roundtrips of 64 B binary data:   0.9s    1.43 MiB/s
5000 roundtrips of 16 KiB binary data:  1.5s    101.33 MiB/s
1000 roundtrips of 128 KiB binary data: 10.1s   24.77 MiB/s
100 roundtrips of 1 MiB binary data:    7.3s    27.53 MiB/s
1 roundtrips of 100 MiB binary data:    8s      24.93 MiB/s
10000 roundtrips of 64 B text data:     0.9s    1.41 MiB/s
5000 roundtrips of 16 KiB text data:    1.6s    96.19 MiB/s
1000 roundtrips of 128 KiB text data:   9.7s    25.65 MiB/s
100 roundtrips of 1 MiB text data:      7.7s    26.09 MiB/s
1 roundtrips of 100 MiB text data:      8.2s    24.29 MiB/s
```
Synchronous zlib appears to provide performance gains of 20% to 40% depending on packet size. If anyone wants to experiment with it and test for memory issues, the code is available in my fork of `ws`.

Cheers!

**Comment by alexhultman** [comment]
> 10000 roundtrips of 64 B binary data:   2.9s    437.94 KiB/s

vs.

> 10000 roundtrips of 64 B binary data:   1.1s    1.11 MiB/s

Is **2.6x** as fast and showcases a huge difference for the most common path; small messages. Most apps send small messages over WS.

Good benchmark

**Comment by lpinca** [comment]
`fast-zlib` is bit hacky but that is not the a problem, we can propose an API to use inflate/deflate streams synchronously in Node.js core.

The problem as I see it is doing this synchronously when are thousands of clients. Yes, it will be slower but doing it asynchronously does not block the event loop.

The benchmark uses only a single client and reflect the results of running this benchmark https://github.com/websockets/ws/issues/1369#issuecomment-415950185. There is no need to use `ws` or transmit data over the network to benchmark this.

One more thing, it doesn't make sense to compress small messages. It is actually counterproductive. This is why `ws` does not compress messages whose size is < 1024 bytes by default.

**Comment by timotejroiko** [comment]
Indeed, i intentionally disabled the threshold to test zlib in all situations, to avoid having mixing compressed and non-compressed results

**Comment by piranna** [comment]
> The problem as I see it is doing this synchronously when there are thousands of clients. Yes, it will be slower but doing it asynchronously does not block the event loop.

What about a flag to left user decide to use them sync or async?

**Comment by lpinca** [comment]
It's an option but it complicates everything :(. The user can already compress the message synchronously before sending it in a WebSocket message. It's not the same but close enough.

**Comment by alexhultman** [comment]
> One more thing, it doesn't make sense to compress small messages. It is actually counterproductive. This is why ws does not compress messages whose size is < 1024 bytes by default.

99% of all WS apps always send messages smaller than 1KB, so this rule is essentially disabling compression altogether. Compressing small messages can be counterproductive without a sliding window, yes, but with a sliding window you're almost guaranteed to get very good compression ratios even for small messages. Some users do consider this valuable and even something small as 60 bytes can get compressed with very high ratios if the data is repeated to a high degree.

> The problem as I see it is doing this synchronously when there are thousands of clients. Yes, it will be slower but doing it asynchronously does not block the event loop.

I made a simple drawing, it does not take into account the delay between signalling the event-loop from the thread-pool to actually getting the signal, which can be many ms of latency. The point is to illustrate the overhead of doing small CPU-tasks on other threads.

![async overhead](https://user-images.githubusercontent.com/3397140/91635259-2ed31200-e9f7-11ea-95d6-542b36947be5.png)

**Comment by lpinca** [comment]
> 99% of all WS apps always send messages smaller than 1KB, so this rule is essentially disabling compression altogether.

What is this based on? Probably true for stock trading apps but 99% of all apps seems a bit too broad. The default value can be changed but:

1. Compression is already disabled by default. I think there is not much need for it when using an efficient binary serialization/deserialization format.
1. User can customize the threshold to fit their needs.

**Comment by alexhultman** [comment]
It's based on observation. It doesn't have to be true, but in my observation so far "IO games", stock trading, currency trading, various completion events, recommendations, chat messages, "status badges", torrent trackers, webrtc signalling servers and the like all deal with less than 1kb per message - often times less than 100 bytes.

Generally people use HTTP for large file transfers and WebSockets for small "events". It doesn't have to be this way, but this is generally true from what I have seen.

**Comment by lpinca** [comment]
I've seen and worked on WS apps used to transfer huge, used to stream audio data, etc. Those apps used messages way bigger than 1 KB (64 KB / 16 KB).

**Comment by alexhultman-2fa** [comment]
Yes some use it like that. Strange. But it doesn't invalidate the observation that the vast majority of use cases don't. And it doesn't matter, really, all that matters is that small message sending is a huge use case with WS. So having an optimized/stable execution path that works well for small messages is important. That's all.

**Comment by mk-pmb** [comment]
I agree, we need an optimized path for applications that are basically signalling or chat, because that's how I will primarily use it. (While still supporting large media transfer as well.)

**Comment by lpinca** [comment]
@mk-pmb to stay on topic, from my experience compressing very small messages (regardless of how compression is done, sync or async)  does not make an app faster but slower.

If you are not happy with `ws` performance when dealing with very small messages you are in the wrong repository because the bottleneck is not `ws` but Node.js `{net,tls}.Socket`, so you should bring this up to Node.js core repository.

**Comment by mk-pmb** [comment]
> compressing very small messages […] does not make an app faster but slower.

Thanks for reminding. I should have expressed myself more clearly. I meant that even in scenarios where network traffic is the bottle neck, where a few bytes saved make a difference, and it's thus acceptable to invest some more CPU cycles, our compression should still try to be economical in how many extra CPU cycles users would need to invest.

**Comment by fungiboletus** [comment]
The discussion about the performances is interesting, but I want to mention that the memory leak is still there with ws@7.3.1 and node 14.9.0. The current implementation may be faster in some uses cases and slower in some others, but it does leak memory.

**Comment by lpinca** [comment]
@fungiboletus I was not able to reproduce the memory fragmentation issue using you code above, see https://github.com/websockets/ws/issues/1369#issuecomment-553831310.

**Comment by KVROHITH098** [comment]
So did we chi the node? And did we pre check the JS implementation of the new node being installed?

**Comment by GITHUBGUY12357999** [comment]
I am going to chi and let u know may I have the link to access the data script for this to do further investigation on my teams side

**Comment by GITHUBGUY12357999** [comment]
Ok So I got a status error while I was looking for the status 


Proof:-- 

21:17:43 quiet period for node-test-commit-arm-fanned is 120 seconds.Finished Build : #43307 of Job : node-test-linter with status : FAILURE at 21:17:45
21:17:45 Finished Build : #42363 of Job : node-test-commit-freebsd with status : FAILURE at 21:17:45
21:17:45 Finished Build : #44176 of Job : node-test-commit-linux with status : FAILURE at 21:17:45
21:19:48 Finished Build : #41618 of Job : node-test-commit-plinux with status : FAILURE at 21:19:48
21:19:48 Finished Build : #23465 of Job : node-test-commit-arm-fanned with status : FAILURE at 21:19:48
21:46:10 Finished Build : #30121 of Job : node-test-commit-linuxone with status : FAILURE at 21:46:10
21:47:15 Finished Build : #164 of Job : node-test-commit-arm-debug with status : UNSTABLE at 21:47:15
21:47:15 Collecting metadata...
21:47:15 Metadata collection done.
21:47:15 Notifying upstream projects of job completion
21:47:15 Finished: FAILURE

**Comment by GITHUBGUY12357999** [comment]
So I have a new way right so what if we test the code first before actually implementing it. 

So can we test it?

Then can we implement it?

**Comment by heroboy** [comment]
Node now has synchronous zlib streams.
https://nodejs.org/docs/latest/api/zlib.html#iterable-compression
